### DISCUSSION TITLE
How to work with link on navbar and CSS?

### DISCUSSION THREAD

**Discussion by Elindo586** [vercel/next.js#36897] [Help] [answered]
OK then, maybe this is the wrong approach about how to do a simple navbar..  if anybody has any inputs it would help.

I am using useEffect and useState to change the CSS of a navbar item given the window.location.pathname.

I am also using link to link the navbar item to a page.

This is just a simple ul  converted into a navbar.

What is the problem?   ---->  The link text isn't changing its CSS (of course not)...  How can I fix this??  ......   Would there be a different approach for Nextjs?

```
<dl
            className={
              pathName === "/services/preventive-maintenance-programs"
                ? "link2 active2"
                : "link2"
            }
          >
            <Link href="/services/preventive-maintenance-programs">
              Programs
            </Link>
          </dl>

```
Complete code here:
https://github.com/Elindo586/nextprojecta/blob/main/pages/components/servicesNav.js

**Comment by raul-saez** [comment]


**Reply by Elindo586** [reply]
@raul-saez   You mean like this? 

 ```
<dl>
<Link href="/services/preventive-maintenance-programs">
           <a className={
              pathName === "/services/preventive-maintenance-programs"
                ? "link2 active2"
                : "link2"
            }  Programs
/>Programs

            </Link>
</dl>
```

Apparently you cannot have more than one link on a link tag.

```
Unhandled Runtime Error
Error: Multiple children were passed to <Link> with `href` of `/services/preventive-maintenance-programs` but only one child is supported https://nextjs.org/docs/messages/link-multiple-children 
Open your browser's console to view the Component stack trace.
```

**Reply by raul-saez** [reply]
No, that css code I shared needs to go in your `general.css` file.

**Reply by Elindo586** [reply]
> No, that css code I shared needs to go in your `general.css` file.

Yes, that is in general.css, but I can't apply it directly to an a tag, because the a tag needs to change CSS depending on the web address that you are viewing.

**Reply by raul-saez** [reply]
You don't need to edit the code on `serviceNav.js` - keep it as you shared in your first example. Simply add the CSS to `general.css`.

**Reply by Elindo586** [reply]
would it change state?.. it doesn't seem to be working.

**Comment by Elindo586** [comment]
I want a dl that would change its CSS, including the link, according to the web address...   I want to apply something like this:

```
function ServicesNav() {
  const [pathName, setPathName] = useState();

  useEffect(() => {
    const url = window.location.pathname;
    setPathName(url);
  });
```

^^^  but the internal text link within the dl isn't taking the changes, but the rest of the dl does...

**Comment by Elindo586** [comment, marked as answer]
Solution...

use passHref at the end of the closing of the first link tag... like this...

Its tricky to access the CSS in Nextjs.

```
<dl>
            <Link href="/services/preventive-maintenance-programs" passHref>
              <a
                className={
                  pathName === "/services/preventive-maintenance-programs"
                    ? "link2 active2"
                    : "link2"
                }
              >
                Programs
              </a>
            </Link>
          </dl>
```

**Reply by Elindo586** [reply]
Can't wait for how interesting this is gonna get working with all the dropdown menus and side effects on the big navbar.

**Reply by icyJoseph** [reply]
This was not really correct, you should use uaeRouter hook, and read pathname, route or asPath from it.

I'd recommended using asPath, with startsWith to check if the pathname is active. Alternatively, pathname works.

Like so:

```jsx

const {pathname} = useRouter()

// Later on

<Link href="/something">
 <a className={pathname === "something" ? "active" : ""}></a>
</Link>
```

More info: https://nextjs.org/docs/api-reference/next/router#router-object

And an actual example from the docs: https://github.com/vercel/next.js/blob/canary/examples/active-class-name/components/ActiveLink.js

**Reply by icyJoseph** [reply]
Your hack seems to work because, iirc, the passHref, in this case is triggering a children re-render, which are otherwise memoised.

The right way to check if a link is active, is by comparing the href with the current pathname, from the Next router.

The Next.js example above is a bit complex but for your use the pathname is sufficient. 

You are using a use-effect without dependencies to keep track of the window location pathname, but that's not necessary, just use `next/router ` which has all the info you need.

**Reply by Elindo586** [reply]
@icyJoseph 

Question... 

By using next/router... then I wouldn't need to use useEffect and useState? .. That would be cool... I like shorter codes.

**Reply by icyJoseph** [reply]
Exactly. Just read from the router object. No need to re-invent the wheel.

Also, I think your solution with useEffect and useState wouldn't work if the component is wrapped with memo, because it's inputs wouldn't change between renders.
